crates/sui-types/src/temporary_store.rs

```diff
@@ -862,13 +867,22 @@ impl<S> TemporaryStore<S> {
         self.deleted.insert(*id, (ctx.clone(), version, kind));
     }
 
-    /// Resets any mutations and deletions recorded in the store.
-    pub fn reset(&mut self) {
+    pub fn drop_writes(&mut self) {
         self.written.clear();
         self.deleted.clear();
         self.events.clear();
     }
 
+    /// Resets any mutations, deletions, and events recorded in the store, as well as any storage costs and
+    /// rebates, then Re-runs gas smashing. Effects on store are now as if we were about to begin execution
+    fn reset(&mut self, sender: SuiAddress, gas: &[ObjectRef], gas_status: &mut SuiGasStatus<'_>) {
+        self.drop_writes();
+        gas_status.reset_storage_cost_and_rebate();
+
+        self.smash_gas(sender, gas)
+            .expect("Gas smashing cannot fail because we checked for sufficient budget up front");
```

Comment:
```suggestion
            .expect("Gas smashing cannot fail because it already succeeded when we did it before on the same `gas`");
```

This is a bad error message by me--there's no gas charging going on in `smash_gas`

Reply:
changed